Fix #510 (miss-matched tag names)

hraftery opened the pull request:
Make `ab_server` lookup tag names based on their entire name, not just their prefix. See single commit message for details.

kyle-github committed:
Prerelease merge into release (#491)

* Make file number optional if input or output file.  Fix #481.

* Added fix for omitted data file number on I or O files. Added support for a numeric subelement.  Fix #479. (#483)

* Fix #486. (#487).   String size change problem.   Fixed with #488.

* Added atomic flag and checks to aid in making all sessions terminate when the library is shutting down. Fix #450.

* plc_tag_set_string() feature additions (#488).   Setting string can change tag data buffer size.  And added flag, allow_field_resize, to suppress change.

* Added msp response size check before adding new requests to msp with allow_packing=1 and fragmented reads disabled (#489)

* Support for Omron to use multi-request packets.

---------

Co-authored-by: Phil Smith <105802449+ptsOSL@users.noreply.github.com>

hraftery committed:
Fix #510 (miss-matched tag names)

Realised that the one-sided length check in `slice_match_bytes` is deliberate - it is used to check just the specified number of bytes in the slice, ignoring what comes after.
Also realised `slice_match_string` is only used by the function in question, so is free to be modified. Gave it a distinguishing name and added the length test so tag names are uniquely identified, instead of just by their prefix.
Also confirmed that `name_len` can be trusted - the protocol says the "Symbolic Segment" includes the length of the name in bytes, and is padded to 16-bit alignment for the following fields only. Ref: https://literature.rockwellautomation.com/idc/groups/literature/documents/pm/1756-pm020_-en-p.pdf

kyle-github commented:
I moved this to the prerelease branch.   That's where all the work is happening right now.

hraftery commented:
> I moved this to the prerelease branch. That's where all the work is happening right now.

Cool. I think the tests are failing because the wrong ref is configured:

https://github.com/libplctag/libplctag/blob/prerelease/.github/workflows/ci.yml#L28

Looks like the ref setting needs to be removed, so the default can take effect. The [default](https://github.com/actions/checkout?tab=readme-ov-file#usage) knows it's a PR, so not to go looking for the source by the branch name. The default seems to be set [here](https://github.com/actions/checkout/blob/main/src/input-helper.ts#L63) but I'm out of my depth.

Note if there are other test issues, I fixed some recently here: https://github.com/libplctag/libplctag/pull/504

Sorry if I'm confusing things all over the place here, I'm a bit lost following all the merges and may not be contributing well!

Lastly, the issue didn't seem to auto link to this PR, so tagging it here: #510

kyle-github commented:
I think I have this fixed now.   I did more brain surgery than you had.  I ended up splitting some of the slice functions into exact and partial matches.

kyle-github commented:
Note that it does not build on Windows yet.   Coming soon.

kyle-github commented:
It builds on Windows.   I am going to skip this PR for now.   I split the match functions in slice.h to have exact and prefix matching.  I use only the exact version in the tag name matches.   I will have a release out within a few days.   Thank you very much for noting this problem!!!!
